### PR TITLE
DOCSP-12159: Data Formats POJOs

### PR DESCRIPTION
**Pull request by ccho-mongodb**
### Issue JIRA link:
https://jira.mongodb.org/browse/DOCSP-12159

### Docs staging link (requires sign-in on MongoDB Corp SSO):
https://docs-mongodborg-staging.corp.mongodb.com/056c933/java/docsworker-xlarge/DOCSP-12159-fundamentals-pojo/fundamentals/data-formats/document-data-format-pojo

### Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Does it render on staging correctly?
- [x] Are all the links working?
- [x] Are the staging links in the PR description updated?

### CODE REVIEW

File: source/fundamentals/data-formats/document-data-format-pojo.txt [hunk 1]
```diff
@@ -13,17 +13,215 @@ Document Data Format: POJOs
 Overview
 --------
 
-- Description; convert between BSON and POJOs  
-- Summary of contents of this doc
+This guide explains how you can store and retrieve MongoDB data using plain
+old Java objects (`POJOs <https://en.wikipedia.org/wiki/Plain_old_Java_object>`__).
+POJOs are classes that do not depend on any other files or frameworks, and are
```
Comment (nathan-contino-mongo) [inline source/fundamentals/data-formats/document-data-format-pojo.txt:18]: Nitpick: "any other" seems vague to me -- does this mean that I can't extend a class in my project to create a POJO if it lives in another file? I would suggest "third party" or "external" as potential alternatives.

File: source/fundamentals/data-formats/document-data-format-pojo.txt [hunk 2]
```diff
@@ -13,17 +13,215 @@ Document Data Format: POJOs
...
+   serialization:
+
+   - **public** getter methods
+   - **public** setter methods
+   - **public** fields
+
+   Also note that:
+
+   - The codec omits static and transient fields in your POJO.
+   - If your field has a corresponding getter method, the codec calls it for
```
Comment (nathan-contino-mongo) [inline source/fundamentals/data-formats/document-data-format-pojo.txt:107]: Suggestion: It might be useful to define what determines a "corresponding getter method". I'd assume that this refers to a method with:
- the name [get|set]<fieldName>
- parameter/return types that match the field's type
I know that some folks have different conventions for accessors and mutators, so we might want to clarify our assumption about what "getter" and "setter" mean.

File: source/fundamentals/data-formats/document-data-format-pojo.txt [hunk 3]
```diff
@@ -13,17 +13,215 @@ Document Data Format: POJOs
...
+   Make sure your POJO properties meet the following requirements for
+   serialization:
+
+   - **public** getter methods
+   - **public** setter methods
+   - **public** fields
+
+   Also note that:
+
+   - The codec omits static and transient fields in your POJO.
```
Comment (nathan-contino-mongo) [inline source/fundamentals/data-formats/document-data-format-pojo.txt:106]: Suggestion: Considering that I didn't remember the definition of transient, it might be helpful to link users to the definition so they aren't stuck wondering if any of their fields meet the definition of "transient". Alternatively, we could omit mention of transient entirely, since the label just means "don't serialize this".

File: source/fundamentals/data-formats/document-data-format-pojo.txt [hunk 4]
```diff
@@ -13,17 +13,215 @@ Document Data Format: POJOs
...
+   - If your field has a corresponding getter method, the codec calls it for
+     the value of the field. Otherwise, it retrieves the value directly from
+     the field.
+   - If your field has a corresponding setter method, the codec calls it to
+     set the value of the field. Otherwise, it sets the value directly on the
+     field.
```
Comment (nathan-contino-mongo) [inline source/fundamentals/data-formats/document-data-format-pojo.txt:112]: These "otherwise," sentences make me wonder: if codecs will get/set values directly on fields, are public getter/setter methods really necessary here? If so, why are they listed above as requirements?

File: source/fundamentals/data-formats/document-data-format-pojo.txt [hunk 5]
```diff
@@ -13,17 +13,215 @@ Document Data Format: POJOs
...
+   }
+
+.. note:: POJO Serialization
+
+   Make sure your POJO properties meet the following requirements for
+   serialization:
+
+   - **public** getter methods
+   - **public** setter methods
+   - **public** fields
```
Comment (nathan-contino-mongo) [inline source/fundamentals/data-formats/document-data-format-pojo.txt:102]: Are public fields a strict requirement for serialization? I was under the impression that if you have public getters and setters, you can set your fields to private to follow best encapsulation practices.

File: source/fundamentals/data-formats/document-data-format-pojo.txt [hunk 6]
```diff
@@ -13,17 +13,215 @@ Document Data Format: POJOs
...
+Configure the Driver for POJOs
+------------------------------
+
+To set up the driver to store and retrieve POJOs, we need to specify:
+
+- A codec provider that contains :doc:`Codecs </fundamentals/data-formats/codecs>`
+  that define how to encode/decode the data between the POJO and MongoDB
+  document, and which POJO classes or packages that the Codecs should apply to.
+- A :java-docs:`CodecRegistry <apidocs/bson/org/bson/codecs/configuration/CodecRegistry.html>`
+  instance that contains the aforementioned information.
```
Comment (nathan-contino-mongo) [inline source/fundamentals/data-formats/document-data-format-pojo.txt:123]: "information" or "codec provider"?

File: source/fundamentals/data-formats/document-data-format-pojo.txt [hunk 7]
```diff
@@ -13,17 +13,215 @@ Document Data Format: POJOs
...
+   to apply the Codecs to any class and its properties.
+
+   .. code-block:: java
+
+      CodecProvider pojoCodecProvider = PojoCodecProvider.builder().automatic(true).build();
+
+.. note::
+
+   Codec providers also contain other objects such as ``ClassModel`` and
+   ``Convention`` instances that further define the serialization behavior.
```
Comment (nathan-contino-mongo) [inline source/fundamentals/data-formats/document-data-format-pojo.txt:146]:
```suggestion
   ``Convention`` instances that further define serialization behavior.
```

File: source/fundamentals/data-formats/document-data-format-pojo.txt [hunk 8]
```diff
@@ -13,17 +13,215 @@ Document Data Format: POJOs
...
+   .. code-block:: java
+
+      // ensure you use the following static imports above your class definition
+      import static com.mongodb.MongoClientSettings.getDefaultCodecRegistry;
+      import static org.bson.codecs.configuration.CodecRegistries.fromProviders;
+      import static org.bson.codecs.configuration.CodecRegistries.fromRegistries;
+
+      // ...
+
+      CodecRegistry pojoCodecRegistry = fromRegistries(getDefaultCodecRegistry(), fromProviders(pojoCodecProvider));
```
Comment (nathan-contino-mongo) [inline source/fundamentals/data-formats/document-data-format-pojo.txt:165]: It might be unclear to users right now why we add the default codec registry before the `pojoCodecProvider`. Consider adding a note or a sentence explaining this bit of magic.

File: source/fundamentals/data-formats/document-data-format-pojo.txt [hunk 9]
```diff
@@ -13,17 +13,215 @@ Document Data Format: POJOs
...
+
+3. Configure our ``MongoClient``, ``MongoDatabase``, or ``MongoCollection``
+   instance to use the Codecs in the ``CodecRegistry``. You only need to
+   configure one of these. In this example, we set it on a ``MongoDatabase``
```
Comment (nathan-contino-mongo) [inline source/fundamentals/data-formats/document-data-format-pojo.txt:169]: Nitpick: "set it" is ambiguous

File: source/fundamentals/data-formats/document-data-format-pojo.txt [hunk 10]
```diff
@@ -13,17 +13,215 @@ Document Data Format: POJOs
...
+   called ``sample_pojos`` using the ``withCodecRegistry()`` method.
+
+   .. code-block:: java
+
+      try (MongoClient mongoClient = MongoClients.create(uri)) {
```
Comment (nathan-contino-mongo) [inline source/fundamentals/data-formats/document-data-format-pojo.txt:174]: Considering how expensive each `MongoClient` is, would we be better off leaving the try-with-resources pattern out, since most end users probably won't use a `MongoClient` like that?

File: source/fundamentals/data-formats/document-data-format-pojo.txt [hunk 11]
```diff
@@ -13,17 +13,215 @@ Document Data Format: POJOs
...
+
+      MongoCollection<Flower> collection = database.getCollection("flowers", Flower.class);
+
+Once you have configured the ``MongoCollection`` instance above, you can:
+
+- save an instance of the POJO to the collection
+- retrieve instances of the POJO from a query on the collection
+
+The code below shows you how you can set data in an instance of a ``Flower``,
+sve it in the collection, and retrieve the data in a ``List`` of ``Flower``
```
Comment (nathan-contino-mongo) [inline source/fundamentals/data-formats/document-data-format-pojo.txt:192]:
```suggestion
save it in the collection, and retrieve the data in a ``List`` of ``Flower``
```

**Review by nathan-contino-mongo — COMMENTED** [review]
Overall I think this does a wonderful job of demonstrating how to use POJOs in the Java driver. Quite a few suggestions and thoughts, but I think this is definitely a solid start.

**Comment by nathan-contino-mongo** [comment]
Additional thought, right after submitting the review. It doesn't look like this page covers these two topics in the JIRA ticket:
-     Nested POJOs
-     Lists
-     Show examples of all the annotations, e.g. BsonProperty, BsonId
Do we plan on covering these elsewhere, or do you just think they complicate the page too much so you've omitted them in the first pass?

**Comment by ccho-mongodb** [comment]
> Additional thought, right after submitting the review. It doesn't look like this page covers these two topics in the JIRA ticket:
> 
> * Nested POJOs
> * Lists
> * Show examples of all the annotations, e.g. BsonProperty, BsonId
> 
> Do we plan on covering these elsewhere, or do you just think they complicate the page too much so you've omitted them in the first pass?

Ticket needs to be updated, thanks for pointing that out. Per our discussions about restructuring the content, Jeff and I agreed upon a new outline, the contents of which are in the last version of this file.

File: source/fundamentals/data-formats/document-data-format-pojo.txt (hunk 1 above)
Comment (ccho-mongodb) [reply source/fundamentals/data-formats/document-data-format-pojo.txt:18]: I guess I wasn't successful in trying to provide a concise description of POJOs. Thanks for the suggestions, I'll try to rework the description.

File: source/fundamentals/data-formats/document-data-format-pojo.txt (hunk 10 above)
Comment (ccho-mongodb) [reply source/fundamentals/data-formats/document-data-format-pojo.txt:174]: Thanks for noting this. I think Jeff decided he preferred this pattern for sample code in our docs (except for when demonstrating MongoClientSettings.builder()), and it's been the pattern throughout our documentation. I think it could be useful to demonstrate efficient code if we had a tutorial on building production apps or making your code more efficient.

File: source/fundamentals/data-formats/document-data-format-pojo.txt (hunk 4 above)
Comment (ccho-mongodb) [reply source/fundamentals/data-formats/document-data-format-pojo.txt:112]: I agree this section is confusing and will rewrite it. I was reusing some information on the existing POJO page but that doesn't seem to explain the whole story either.

File: source/fundamentals/data-formats/document-data-format-pojo.txt (hunk 3 above)
Comment (ccho-mongodb) [reply source/fundamentals/data-formats/document-data-format-pojo.txt:106]: Decided to format these as keywords for clarity.

File: source/fundamentals/data-formats/document-data-format-pojo.txt (hunk 6 above)
Comment (ccho-mongodb) [reply source/fundamentals/data-formats/document-data-format-pojo.txt:123]: Updating to be more specific.

File: source/fundamentals/data-formats/document-data-format-pojo.txt (hunk 5 above)
Comment (ccho-mongodb) [reply source/fundamentals/data-formats/document-data-format-pojo.txt]: Updated this section as mentioned in prior comment.

File: source/fundamentals/data-formats/document-data-format-pojo.txt (hunk 8 above)
Comment (ccho-mongodb) [reply source/fundamentals/data-formats/document-data-format-pojo.txt:168]: Whoops, forgot to address this item, will fix shortly

File: source/fundamentals/data-formats/document-data-format-pojo.txt [hunk 12]
```diff
@@ -13,17 +13,227 @@ Document Data Format: POJOs
...
+If you are creating your own POJO for storing and retrieving data in MongoDB,
+make sure to follow the following guidelines:
+
+- The POJO class should not implement interfaces or extend classes from a
+  framework.
+- Include all the fields for which you want to store and retrieve data;
+  make sure they are not ``static`` or ``transient``.
+- If you include public getter or setter methods using the
+  `JavaBean naming conventions <https://docs.oracle.com/javase/tutorial/javabeans/writing/properties.html>`__
+  in your POJO, they get called when serializing or deserializing data.  If
```
Comment (nathan-contino-mongo) [inline source/fundamentals/data-formats/document-data-format-pojo.txt:110]: nit: "the get called" sounds informal compared to the rest of this document, and could be confusing since it's close to multiple occurrences of the word "getters". I'd suggest rewording to something like "the driver calls those methods when serializing or deserializing data".

Comment (ccho-mongodb) [reply source/fundamentals/data-formats/document-data-format-pojo.txt:110]: Thanks for spotting this, and good suggestion.

File: source/fundamentals/data-formats/document-data-format-pojo.txt [hunk 13]
```diff
@@ -13,17 +13,227 @@ Document Data Format: POJOs
...
+
+- The ``PojoCodecProvider``, a codec provider that includes
+  :doc:`Codecs </fundamentals/data-formats/codecs>` that define how to
+  encode/decode the data between the POJO and MongoDB document, and which
+  POJO classes or packages that the Codecs should apply to.
+- A :java-docs:`CodecRegistry <apidocs/bson/org/bson/codecs/configuration/CodecRegistry.html>`
+  instance that contains the codec provider and other related information.
+- A ``MongoClient``, ``MongoDatabase``, or ``MongoCollection`` instance
+  configured to use the ``CodecRegistry``.
+- A ``MongoCollection`` instance configured to cast to your POJO document
```
Comment (nathan-contino-mongo) [inline source/fundamentals/data-formats/document-data-format-pojo.txt:128]: nit/question: is "cast to" the correct verbiage here? Or would it be better to word this something like "serialize to"? (personally, I do think "cast" might be the most understandable verb here, but I worry that it's not strictly accurate)

File: source/fundamentals/data-formats/document-data-format-pojo.txt [hunk 14]
```diff
@@ -13,17 +13,227 @@ Document Data Format: POJOs
...
+
+
+#. Add the ``PojoCodecProvider`` instance to a ``CodecRegistry``. The
+   ``CodecRegistry`` allows you to specify one or more codec providers to
+   encode the POJO data. In this example, we call the following methods:
+
+   - :java-docs:`fromRegistries() </apidocs/bson/org/bson/codecs/configuration/CodecRegistries.html#fromRegistries(org.bson.codecs.configuration.CodecRegistry...)>`
+     to combine multiple ``CodecRegistry`` instances into a single one
+   - :java-docs:`getDefaultCodecRegistry() <apidocs/mongodb-driver-core/com/mongodb/MongoClientSettings.html?is-external=true#getDefaultCodecRegistry()>`
+     to retrieve a CodecRegistry instance from a list of codec providers
```
Comment (nathan-contino-mongo) [inline source/fundamentals/data-formats/document-data-format-pojo.txt:161]:
```suggestion
     to retrieve a ``CodecRegistry`` instance from a list of codec providers
```

Missing a monospace here.

File: source/fundamentals/data-formats/document-data-format-pojo.txt [hunk 15]
```diff
@@ -13,17 +13,227 @@ Document Data Format: POJOs
...
+   - :java-docs:`fromProviders() <apidocs/bson/org/bson/codecs/configuration/CodecRegistries.html#fromProviders(org.bson.codecs.configuration.CodecProvider...)>`
+     to create a ``CodecRegistry`` instance from the ``PojoCodecProvider``.
```
Comment (nathan-contino-mongo) [inline source/fundamentals/data-formats/document-data-format-pojo.txt:163]: nit: bullet point consistency, should either remove the period at the end of the bullet point or add one to the other two bullet points.

File: source/fundamentals/data-formats/document-data-format-pojo.txt [hunk 16]
```diff
@@ -13,17 +13,227 @@ Document Data Format: POJOs
...
+      import static org.bson.codecs.configuration.CodecRegistries.fromRegistries;
+
+      // ...
+
+      CodecRegistry pojoCodecRegistry = fromRegistries(getDefaultCodecRegistry(), fromProviders(pojoCodecProvider));
+
+#. Configure our ``MongoClient``, ``MongoDatabase``, or ``MongoCollection``
+   instance to use the Codecs in the ``CodecRegistry``. You only need to
+   configure one of these. In this example, we set the ``CodecRegistry`` on a
+    ``MongoDatabase`` called ``sample_pojos`` using the ``withCodecRegistry()``
```
Comment (nathan-contino-mongo) [inline source/fundamentals/data-formats/document-data-format-pojo.txt:181]: I believe there's an extra space at the beginning of this line that's forcing a newline here in the staged output.

**Review by nathan-contino-mongo — APPROVED** [review]
LGTM overall, a couple of small nits and typos but overall some really high quality content. Nice job!

File: source/fundamentals/data-formats/document-data-format-pojo.txt (hunk 13 above)
Comment (ccho-mongodb) [reply source/fundamentals/data-formats/document-data-format-pojo.txt:128]: I think cast is the verbiage used in the api docs. For example https://mongodb.github.io/mongo-java-driver/4.1/apidocs/bson/org/bson/Document.html

File: source/fundamentals/data-formats/document-data-format-pojo.txt (hunk 15 above)
Comment (ccho-mongodb) [reply source/fundamentals/data-formats/document-data-format-pojo.txt:163]: Whoops, missed that, thanks.

File: source/fundamentals/data-formats/document-data-format-pojo.txt [hunk 17]
```diff
@@ -13,17 +13,226 @@ Document Data Format: POJOs
...
+make sure to follow the following guidelines:
+
+- The POJO class should not implement interfaces or extend classes from a
+  framework.
+- Include all the fields for which you want to store and retrieve data;
+  make sure they are not ``static`` or ``transient``.
+- If you include public getter or setter methods using the
+  `JavaBean naming conventions <https://docs.oracle.com/javase/tutorial/javabeans/writing/properties.html>`__
+  in your POJO, the driver calls them when serializing or deserializing data.
+  If you omit the getter or setter methods for a property field, the driver
```
Comment (rozza) [inline source/fundamentals/data-formats/document-data-format-pojo.txt:111]: Only if the property field is public. Users have to use the `Conventions.SET_PRIVATE_FIELDS_CONVENTION` to allow private fields to be accessed.

Suggest change to `a public property field`

**Review by rozza — CHANGES_REQUESTED** [review]
One minor nit but once thats fixed LGTM

File: source/fundamentals/data-formats/document-data-format-pojo.txt (hunk 17 above)
Comment (ccho-mongodb) [reply source/fundamentals/data-formats/document-data-format-pojo.txt]: Updated with your suggestion in https://github.com/mongodb/docs-java/pull/38/commits/50756bf2baccafa6720e4d65b2ee839f545a2265

**Review by rozza — APPROVED** [review]
LGTM

**Comment by ccho-mongodb** [comment]
> LGTM

Thanks. @jyemin can you review when you get a chance? Thanks!

File: source/fundamentals/data-formats/document-data-format-pojo.txt [hunk 18]
```diff
@@ -13,17 +13,226 @@ Document Data Format: POJOs
...
+
+- The ``PojoCodecProvider``, a codec provider that includes
+  :doc:`Codecs </fundamentals/data-formats/codecs>` that define how to
+  encode/decode the data between the POJO and MongoDB document, and which
+  POJO classes or packages that the Codecs should apply to.
+- A :java-docs:`CodecRegistry <apidocs/bson/org/bson/codecs/configuration/CodecRegistry.html>`
+  instance that contains the codec provider and other related information.
+- A ``MongoClient``, ``MongoDatabase``, or ``MongoCollection`` instance
+  configured to use the ``CodecRegistry``.
+- A ``MongoCollection`` instance configured to cast to your POJO document
```
Comment (jyemin) [inline source/fundamentals/data-formats/document-data-format-pojo.txt:127]: Technically, since it's using generics, it's not casting (or at least, the casts are inserted by the compiler).  Better to say: 

> A MongoCollection instance created with the POJO document class bound to the `TDocument` generic type.

File: source/fundamentals/data-formats/document-data-format-pojo.txt [hunk 19]
```diff
@@ -13,17 +13,226 @@ Document Data Format: POJOs
...
+   method.
+
+   .. code-block:: java
+
+      try (MongoClient mongoClient = MongoClients.create(uri)) {
+          MongoDatabase database = mongoClient.getDatabase("sample_pojos").withCodecRegistry(pojoCodecRegistry);
+          // ...
+      }
+
+#. Configure your ``MongoCollection`` instance to cast to your POJO class when
```
Comment (jyemin) [inline source/fundamentals/data-formats/document-data-format-pojo.txt:190]: See above note about casting

File: source/fundamentals/data-formats/document-data-format-pojo.txt [hunk 20]
```diff
@@ -13,17 +13,226 @@ Document Data Format: POJOs
...
+
+   Flower flower = new Flower("rose", false, 25.4f, Arrays.asList(new String[] {"red", "green"}));
+
+   // insert the instance
+   collection.insertOne(flower);
+
+   // return all documents in the collection
+   List<Flower> flowers = new ArrayList<Flower>();
+
+   MongoCursor<Flower> cursor = collection.find().iterator();
```
Comment (jyemin) [inline source/fundamentals/data-formats/document-data-format-pojo.txt:216]: Use either toArray method on FindIterable, or if you want to show the loop, use try-with-resources statement

**Review by jyemin — CHANGES_REQUESTED** [review]

File: source/fundamentals/data-formats/document-data-format-pojo.txt (hunk 19 above)
Comment (ccho-mongodb) [reply source/fundamentals/data-formats/document-data-format-pojo.txt]: Simplified the text a bit since the generic type was explained earlier, let me know what you think:

> Pass your POJO class to your call to ``getCollection()`` as the document class parameter and specify it as the type argument of your ``MongoCollection`` instance as follows:

File: source/fundamentals/data-formats/document-data-format-pojo.txt (hunk 20 above)
Comment (ccho-mongodb) [reply source/fundamentals/data-formats/document-data-format-pojo.txt]: I updated this to use the `into()` method on the `FindIterable`. Is that what you meant by `toArray`? I wasn't able to find a `toArray()` method on `MongoIterable`.

**Review by jyemin — APPROVED** [review]
